Update cabal.haskell-ci to only run CI on main or PRs targeting main

Mikah-Kainen opened the pull request:
This PR updates `cabal.haskell-ci` to run CI on pushes to main and PRs that target main.

I wasn't able to find a way to configure `haskell-ci` to generate `.yml` files that run CI on all PRs, but if we ever have a PR targeting a branch, we can modify the `.yml` files locally for testing.

The changes to `haskell-ci.yml` came from running `haskell-ci regenerate`. I used a newer version of `haskell-ci`, which introduced some additional changes unrelated to this PR.

Fixes https://github.com/gibbon-compiler/gibbon/issues/297

Mikah-Kainen commented:
Thanks for the detailed feedback! I think I solved the first issue by removing the unused dependencies, but if it turns out we need them for something after all, I will go with your solution of adding
```
error-unused-packages: False
```
to `cabal.haskell-ci`. In general, I think the extra checks are good, so where possible, I will try to keep them enabled and resolve the warnings that trigger them.

I'll tackle the missing-fields initialization tomorrow. It seems like there might be some other problems from `Pattern match(es) are non-exhaustive` and `Defined but not used` (see [here](https://github.com/gibbon-compiler/gibbon/actions/runs/17995226990/job/51193116727?pr=299#step:21:741)), which I will also try to work through.

ulysses4ever commented:
Sorry, I should have explained more. We do need the lower bound on `hashable` otherwise some tests become non-deterministic (see https://github.com/gibbon-compiler/gibbon/pull/292#discussion_r2348850923 and https://github.com/gibbon-compiler/gibbon/pull/295). And there's no good reason to specify this bound without getting the warning (see https://github.com/haskell/cabal/issues/11053).

> Defined but not used, This binding shadows the existing binding and Pattern match(es) are non-exhaustive

the first two are mere warnings, so they should not affect the build. The last one (Pattern match(es) are non-exhaustive) is a warning promoted to error recently (https://github.com/haskell-CI/haskell-ci/pull/790). It'll have to be fixed. Note that I made a pass over warnings in a separate PR https://github.com/gibbon-compiler/gibbon/pull/294 (although my PR addresses only part of them). You risk to duplicate that work, so maybe we should merge that first.

I'm sorry it's taking more effort than anticipated: haskell-CI's update policies are hectic. I didn't realise how much trickier it will be after the update...

Mikah-Kainen commented:
Makes sense, thank you! I’ll take a look at your PR and add back the `hashable` dependency. Do we also need `uuid`?

Mikah-Kainen commented:
I've removed a fair amount of the non-exhaustive uni-patterns and added `{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}` to the rest! Once CI is green, I'll mark this PR ready for review.

ulysses4ever commented:
now the test suite: it also has unused packages:

```
Building test suite 'test-gibbon' for gibbon-0.3...

<no location info>: error: [-Wunused-packages, -Werror=unused-packages]
    The following packages were specified via -package or -package-id flags,
    but were not needed for compilation:
      - bytestring-0.11.5.1 (exposed by flag -package-id bytestring-0.11.5.1)
      - directory-1.3.7.1 (exposed by flag -package-id directory-1.3.7.1)
      - filepath-1.4.2.2 (exposed by flag -package-id filepath-1.4.2.2)
      - pretty-1.1.3.6 (exposed by flag -package-id pretty-1.1.3.6)
      - process-1.6.17.0 (exposed by flag -package-id process-1.6.17.0)
      - srcloc-0.6.0.1 (exposed by flag -package-id srcloc-0.6.0.1-6a5ed5348e24b559ae8413eb8c5a6b68f6e9dbf5d7a38338069c327b446afd8b)
      - text-2.0.2 (exposed by flag -package-id text-2.0.2)
      - time-1.12.2 (exposed by flag -package-id time-1.12.2)
```

Mikah-Kainen commented:
Sounds good! I apologize for the delay.

Mikah-Kainen commented:
Will do!

Mikah-Kainen commented:
I'm closing this PR in favor of https://github.com/gibbon-compiler/gibbon/pull/310, which will include all the CI updates of this PR without the extra bloat. I will open another PR for the other changes later, and link it here where I do.
